Initial commit to System.Net.HttpListener

Priya91 opened the pull request:
This gets @ericeil PR #13287 in. Fixed merged conflicts.

cc @stephentoub @davidsh @CIPop @geoffkizer @karelz

stephentoub commented on src/Common/src/Interop/Windows/HttpApi/Interop.HttpApi.cs:
```diff
@@ -0,0 +1,1017 @@
+
+internal static partial class Interop
+{
+    internal unsafe static partial class HttpApi
```
Can we scope the unsafe to be only where it's needed, rather than the whole class?

Priya91 replied:
fixed.

stephentoub commented on src/Common/src/Interop/Windows/HttpApi/Interop.HttpApi.cs:
```diff
@@ -0,0 +1,1017 @@
+        internal struct HTTP_KNOWN_HEADER
+        {
+            internal ushort RawValueLength;
+            internal sbyte* pRawValue;
```
Why sbyte rather than byte? (Same question for the other occurrences below.)

stephentoub commented on src/Common/src/Interop/Windows/HttpApi/Interop.HttpApi.cs:
```diff
@@ -0,0 +1,1017 @@
+        [DllImport(Libraries.HttpApi, SetLastError = true)]
+        internal static extern uint HttpCloseServerSession(ulong serverSessionId);
+
+        internal class SafeLocalFreeChannelBinding : ChannelBinding
```
Nit: sealed?

Priya91 replied:
fixed.

stephentoub commented on src/Common/src/Interop/Windows/HttpApi/Interop.HttpApi.cs:
```diff
@@ -0,0 +1,1017 @@
+            {
+                SafeLocalFreeChannelBinding result;
+
+                result = LocalAllocChannelBinding(LMEM_FIXED, (UIntPtr)cb);
```
Nit: the declaration and initialization can be combined

stephentoub commented on src/Common/src/Interop/Windows/HttpApi/Interop.HttpApi.cs:
```diff
@@ -0,0 +1,1017 @@
+        }
+
+        [DllImport(Interop.Libraries.Heap, EntryPoint = "LocalAlloc", SetLastError = true)]
+        internal static extern SafeLocalFreeChannelBinding LocalAllocChannelBinding(int uFlags, UIntPtr sizetdwBytes);
```
Nit: Why not just call the method LocalAlloc rather than calling it LocalAllocChannelBinding and using EntryPoint to change its name?

Priya91 replied:
Yeah makes sense, it was copied from Desktop code.

stephentoub commented on src/Common/src/Interop/Windows/HttpApi/Interop.HttpApi.cs:
```diff
@@ -0,0 +1,1017 @@
+            //}
+
+            //NetEventSource.Exit(null);
+            //return endpoint;
```
What's the plan for all of this commented out code?  Is there a TODO to follow-up on this?

Priya91 replied:
Yeah, am working on it.

stephentoub commented on src/Common/src/Interop/Windows/HttpApi/Interop.HttpApi.cs:
```diff
@@ -0,0 +1,1017 @@
+            //}
+
+            //NetEventSource.Exit(null);
+            //return endpoint;
```
Same comment/question

stephentoub commented on src/Common/src/Interop/Windows/sspicli/SafeDeleteContext.cs:
```diff
@@ -0,0 +1,55 @@
+
+namespace System.Net.Security
+{
+    //
```
How does this SafeDeleteContext differ from the one we already have in the repo?
[https://github.com/dotnet/corefx/blob/master/src/Common/src/System/Net/Security/Unix/SafeDeleteContext.cs](https://github.com/dotnet/corefx/blob/master/src/Common/src/System/Net/Security/Unix/SafeDeleteContext.cs)

stephentoub replied:
Oh, I see, you just separated it out of SecuritySafeHandles.cs. Ok.

---
In reply to: [87213631](https://github.com/dotnet/corefx/pull/13502#discussion_r87213631) [](ancestors = 87213631)

stephentoub commented on src/System.Net.HttpListener/src/System/Net/AsyncRequestContext.cs:
```diff
@@ -0,0 +1,80 @@
+
+namespace System.Net
+{
+    internal unsafe class AsyncRequestContext : RequestContextBase
```
Nit: sealed?

Priya91 replied:
fixed.

stephentoub commented on src/System.Net.HttpListener/src/System/Net/AsyncRequestContext.cs:
```diff
@@ -0,0 +1,80 @@
+    {
+        private NativeOverlapped* _nativeOverlapped;
+        private ThreadPoolBoundHandle _boundHandle;
+        private ListenerAsyncResult _result;
```
Nit: readonly

Priya91 replied:
fixed.

stephentoub commented on src/System.Net.HttpListener/src/System/Net/AsyncRequestContext.cs:
```diff
@@ -0,0 +1,80 @@
+            if (_nativeOverlapped != null)
+            {
+                Debug.Assert(!disposing, "AsyncRequestContext::Dispose()|Must call ReleasePins() before calling Dispose().");
+                if (!Environment.HasShutdownStarted || disposing)
```
>!Environment.HasShutdownStarted ||  [](start = 20, length = 35)

Do we know why this check is here?

Priya91 replied:
@CIPop Any idea?

stephentoub commented on src/System.Net.HttpListener/src/System/Net/CaseInsensitiveAsciiComparer.cs:
```diff
@@ -0,0 +1,131 @@
+
+namespace System.Net
+{
+    internal class CaseInsensitiveAsciiComparer : IEqualityComparer, IComparer
```
How does this differ from the CaseInsensitiveAscii comparer already in the repo?
[https://github.com/dotnet/corefx/blob/master/src/System.Net.WebHeaderCollection/src/System/Net/WebHeaderCollection.cs#L496](https://github.com/dotnet/corefx/blob/master/src/System.Net.WebHeaderCollection/src/System/Net/WebHeaderCollection.cs#L496)
If possible, we should combine them to avoid duplication.

Priya91 replied:
Yeah, confirmed it's the same. Will reuse.

stephentoub commented on src/System.Net.HttpListener/src/System/Net/CookieExtensions.cs:
```diff
@@ -0,0 +1,45 @@
+                s_toServerStringFunc = (Func<Cookie, string>)typeof(Cookie).GetMethod("ToServerString").CreateDelegate(typeof(Func<Cookie, string>));
+            }
+            return s_toServerStringFunc(cookie);
+        }
```
@weshaggard, FYI on another reflection case to access internals in another assembly.

Priya91 replied:
Created issue #13607 to track.

stephentoub commented on src/System.Net.HttpListener/src/System/Net/HttpListener.cs:
```diff
@@ -0,0 +1,37 @@
+            //}
+            finally
+            {
+                //if (NetEventSource.IsEnabled) NetEventSource.Exit(NetEventSource.ComponentType.HttpListener, this, "Close", "");
```
All of this tracing should be fixed and uncommented.

stephentoub commented on src/System.Net.HttpListener/src/System/Net/HttpListener.Windows.cs:
```diff
@@ -0,0 +1,2135 @@
+        private string _realm;
+        private SafeHandle _requestQueueHandle;
+        private ThreadPoolBoundHandle _requestQueueBoundHandle;
+        private volatile State _state; // m_State is set only within lock blocks, but often read outside locks. 
```
Nit: wrong name

Priya91 replied:
fixed.

stephentoub commented on src/System.Net.HttpListener/src/System/Net/HttpListener.Windows.cs:
```diff
@@ -0,0 +1,2135 @@
+        private HttpListenerTimeoutManager _timeoutManager;
+        private bool _V2Initialized;
+
+        private Hashtable _disconnectResults;         // ulong -> DisconnectAsyncResult
```
>Hashtable _disconnectResults;         // ulong -> DisconnectAsyncResult [](start = 16, length = 71)

Should this be a ```Dictionary<ulong, DisconnectAsyncResult>``` then?

stephentoub replied:
>Hashtable _disconnectResults;         // ulong -> DisconnectAsyncResult [](start = 16, length = 71)

Nit: _uriPrefixes

Priya91 replied:
fixed.

stephentoub commented on src/System.Net.HttpListener/src/System/Net/HttpListener.Windows.cs:
```diff
@@ -0,0 +1,2135 @@
+        private bool _V2Initialized;
+
+        private Hashtable _disconnectResults;         // ulong -> DisconnectAsyncResult
+        private object _internalLock;
```
readonly

Priya91 replied:
fixed.

stephentoub commented on src/System.Net.HttpListener/src/System/Net/HttpListener.Windows.cs:
```diff
@@ -0,0 +1,2135 @@
+                CheckDisposed();
+                if (value == null)
+                {
+                    throw new ArgumentNullException();
```
nameof(value)?

Priya91 replied:
fixed.

stephentoub commented on src/System.Net.HttpListener/src/System/Net/HttpListener.Windows.cs:
```diff
@@ -0,0 +1,2135 @@
+                        i++;
+                    }
+                }
+                //GlobalLog.Print("HttpListener#" + LoggingHash.HashString(this) + "::AddPrefix() mapped uriPrefix:" + uriPrefix + " to registeredPrefix:" + registeredPrefix);
```
Logging needs to be uncommented/fixed.

stephentoub commented on src/System.Net.HttpListener/src/System/Net/HttpListener.Windows.cs:
```diff
@@ -0,0 +1,2135 @@
+
+        internal bool RemovePrefix(string uriPrefix)
+        {
+            //if (NetEventSource.IsEnabled) NetEventSource.Enter(NetEventSource.ComponentType.HttpListener, this, "RemovePrefix", "uriPrefix:" + uriPrefix);
```
All of this logging needs to be fixed/uncommented... will stop commenting on this.

stephentoub commented on src/System.Net.HttpListener/src/System/Net/HttpListener.Windows.cs:
```diff
@@ -0,0 +1,2135 @@
+                //GlobalLog.Print("HttpListener#" + LoggingHash.HashString(this) + "::RemovePrefix() uriPrefix:" + uriPrefix);
+                if (uriPrefix == null)
+                {
+                    throw new ArgumentNullException("uriPrefix");
```
nameof(uriPrefix)

stephentoub commented on src/System.Net.HttpListener/src/System/Net/HttpListener.Windows.cs:
```diff
@@ -0,0 +1,2135 @@
+                if (_serverSessionHandle != null)
+                {
+                    _serverSessionHandle.Dispose();
+                }
```
Nit: could be simplified to ```_serverSessionHandle?.Dispose();```

Priya91 replied:
fixed.

stephentoub commented on src/System.Net.HttpListener/src/System/Net/HttpListener.Windows.cs:
```diff
@@ -0,0 +1,2135 @@
+            {
+                return false;
+            }
+            return true;
```
Nit: ```return statusCode != Interop.HttpApi.ERROR_NOT_FOUND;```?

Priya91 replied:
fixed.

stephentoub commented on src/System.Net.HttpListener/src/System/Net/HttpListener.Windows.cs:
```diff
@@ -0,0 +1,2135 @@
+                CheckDisposed();
+                if (asyncResult == null)
+                {
+                    throw new ArgumentNullException("asyncResult");
```
nameof(asyncResult)

Priya91 replied:
fixed.

stephentoub commented on src/System.Net.HttpListener/src/System/Net/HttpListener.Windows.cs:
```diff
@@ -0,0 +1,2135 @@
+                ListenerAsyncResult castedAsyncResult = asyncResult as ListenerAsyncResult;
+                if (castedAsyncResult == null || castedAsyncResult.AsyncObject != this)
+                {
+                    throw new ArgumentException(SR.net_io_invalidasyncresult, "asyncResult");
```
nameof(asyncResult)

stephentoub commented on src/System.Net.HttpListener/src/System/Net/HttpListener.Windows.cs:
```diff
@@ -0,0 +1,2135 @@
+
+        public Task<HttpListenerContext> GetContextAsync()
+        {
+            return Task<HttpListenerContext>.Factory.FromAsync(BeginGetContext, EndGetContext, null);
```
This would be better as:
```C#
return Task.Factory.FromAsync<HttpListenerContext>(
    (callback, state) => ((HttpListener)state).BeginGetContext(callback, state),
    iar => ((HttpListener)iar.AsyncState).EndGetContext(iar),
    this);
```
to avoid the begin and end delegate allocations on each invocation.

stephentoub commented on src/System.Net.HttpListener/src/System/Net/HttpListener.Windows.cs:
```diff
@@ -0,0 +1,2135 @@
+            {
+                if (authorizationHeader == null)
+                {
+                    WindowsPrincipal principal = disconnectResult == null ? null : disconnectResult.AuthenticatedConnection;
```
>disconnectResult == null ? null : disconnectResult.AuthenticatedConnection [](start = 49, length = 74)

Could just be ```disconnectResult?.AuthenticatedConnection```.

stephentoub commented on src/System.Net.HttpListener/src/System/Net/HttpListener.Windows.cs:
```diff
@@ -0,0 +1,2135 @@
+                        // Save digest context in digest cache, we may need it later because of
+                        // subsequest responses to the same req on the same/diff connection
+
+                        if ((authenticationScheme
```
Nit: ```nameof(Interop.HttpApi.HTTP_REQUEST_CHANNEL_BIND_STATUS.ChannelToken)``` ?

stephentoub replied:
Nit: ```nameof(Interop.HttpApi.HTTP_REQUEST_CHANNEL_BIND_STATUS.ChannelTokenSize)```?

stephentoub replied:
Can any of these ArrayLists be changed to ```List<T>```?

stephentoub replied:
Nit: sealed?

stephentoub commented on src/System.Net.HttpListener/src/System/Net/HttpListenerBasicIdentity.cs:
```diff
@@ -0,0 +1,27 @@
+            {
+                return _password;
+            }
+        }
```
Nit: could be ```public virtual string Password { get; }```, with the ctor assigning to the property, and then you can get rid of the _password field.  If you choose not to do that, _password can be made readonly.

Priya91 replied:
fixed.

stephentoub commented on src/System.Net.HttpListener/src/System/Net/HttpListenerContext.Unix.cs:
```diff
@@ -0,0 +1,64 @@
+using System.Threading.Tasks;
+
+namespace System.Net
+{
```
Please add a TODO comment here with the associated issue number.

Priya91 replied:
fixed.

stephentoub commented on src/System.Net.HttpListener/src/System/Net/HttpListenerContext.Windows.cs:
```diff
@@ -0,0 +1,264 @@
+            {
+                _authenticationSchemes = value;
+            }
+        }
```
Nit: properties/fields like this could be changed to use auto props, e.g. ```internal AuthenticationSchemes AuthenticationSchemes { get; set; }```

Priya91 replied:
Done.

stephentoub commented on src/System.Net.HttpListener/src/System/Net/HttpListenerContext.Windows.cs:
```diff
@@ -0,0 +1,264 @@
+                if (_response != null)
+                {
+                    _response.Close();
+                }
```
Nit: ```_response?.Close();```

Priya91 replied:
done.

stephentoub commented on src/System.Net.HttpListener/src/System/Net/HttpListenerContext.Windows.cs:
```diff
@@ -0,0 +1,264 @@
+                }
+                finally
+                {
+                    IDisposable user = _user == null ? null : _user.Identity as IDisposable;
```
>_user == null ? null : _user.Identity as IDisposable; [](start = 39, length = 53)

Nit: ```_user?.Identity as IDisposable```

Priya91 replied:
done.

stephentoub commented on src/System.Net.HttpListener/src/System/Net/HttpListenerContext.Windows.cs:
```diff
@@ -0,0 +1,264 @@
+                if (user != null)
+                {
+                    user.Dispose();
+                }
```
Nit: could be:
```C#
(_user?.Identity as IDisposable)?.Dispose();
```

Priya91 replied:
done.

stephentoub commented on src/System.Net.HttpListener/src/System/Net/HttpListenerContext.Windows.cs:
```diff
@@ -0,0 +1,264 @@
+        internal static void CancelRequest(SafeHandle requestQueueHandle, ulong requestId)
+        {
+            // It is safe to ignore the return value on a cancel operation because the connection is being closed
+            uint statusCode = Interop.HttpApi.HttpCancelHttpRequest(requestQueueHandle, requestId,
```
>uint statusCode =  [](start = 12, length = 18)

Nit: unused?

Priya91 replied:
done.

stephentoub commented on src/System.Net.HttpListener/src/System/Net/HttpListenerPrefixCollection.cs:
```diff
@@ -0,0 +1,149 @@
+
+namespace System.Net
+{
+    internal class ListenerPrefixEnumerator : IEnumerator<string>
```
Nit: sealed?

stephentoub commented on src/System.Net.HttpListener/src/System/Net/HttpListenerPrefixCollection.cs:
```diff
@@ -0,0 +1,149 @@
+
+    public class HttpListenerPrefixCollection : ICollection<string>
+    {
+        private HttpListener _httpListener;
```
Nit: readonly?

Priya91 replied:
done

stephentoub commented on src/System.Net.HttpListener/src/System/Net/HttpListenerPrefixCollection.cs:
```diff
@@ -0,0 +1,149 @@
+            _httpListener.CheckDisposed();
+            if (Count > array.Length)
+            {
+                throw new ArgumentOutOfRangeException("array", SR.net_array_too_small);
```
nameof(array)

stephentoub commented on src/System.Net.HttpListener/src/System/Net/HttpListenerPrefixCollection.cs:
```diff
@@ -0,0 +1,149 @@
+            }
+            if (offset + Count > array.Length)
+            {
+                throw new ArgumentOutOfRangeException("offset");
```
nameof(offset)

stephentoub commented on src/System.Net.HttpListener/src/System/Net/HttpListenerPrefixCollection.cs:
```diff
@@ -0,0 +1,149 @@
+            _httpListener.CheckDisposed();
+            if (Count > array.Length)
+            {
+                throw new ArgumentOutOfRangeException("array", SR.net_array_too_small);
```
nameof(array)

stephentoub commented on src/System.Net.HttpListener/src/System/Net/HttpListenerPrefixCollection.cs:
```diff
@@ -0,0 +1,149 @@
+            }
+            if (offset + Count > array.Length)
+            {
+                throw new ArgumentOutOfRangeException("offset");
```
nameof(offset)

stephentoub commented on src/System.Net.HttpListener/src/System/Net/HttpListenerRequest.Unix.cs:
```diff
@@ -0,0 +1,267 @@
+using System.Threading.Tasks;
+
+namespace System.Net
+{
```
As with elsewhere, please add a TODO comment here

stephentoub commented on src/System.Net.HttpListener/src/System/Net/HttpListenerRequest.Windows.cs:
```diff
@@ -0,0 +1,1246 @@
+
+                foreach (string upgrade in this.Headers.GetValues(HttpKnownHeaderNames.Upgrade))
+                {
+                    if (string.Compare(upgrade, WebSocketHelpers.WebSocketUpgradeToken, StringComparison.OrdinalIgnoreCase) == 0)
```
string.Equals

Priya91 replied:
done.

stephentoub commented on src/System.Net.HttpListener/src/System/Net/HttpListenerRequest.Windows.cs:
```diff
@@ -0,0 +1,1246 @@
+            {
+                if (asyncResult == null)
+                {
+                    throw new ArgumentNullException("asyncResult");
```
nameof(asyncResult)

stephentoub commented on src/System.Net.HttpListener/src/System/Net/HttpListenerRequest.Windows.cs:
```diff
@@ -0,0 +1,1246 @@
+                ListenerClientCertAsyncResult clientCertAsyncResult = asyncResult as ListenerClientCertAsyncResult;
+                if (clientCertAsyncResult == null || clientCertAsyncResult.AsyncObject != this)
+                {
+                    throw new ArgumentException(SR.net_io_invalidasyncresult, "asyncResult");
```
nameof(asyncResult)

stephentoub commented on src/System.Net.HttpListener/src/System/Net/HttpListenerRequest.Windows.cs:
```diff
@@ -0,0 +1,1246 @@
+                }
+                if (clientCertAsyncResult.EndCalled)
+                {
+                    throw new InvalidOperationException(SR.Format(SR.net_io_invalidendcall, "EndGetClientCertificate"));
```
nameof(EndGetClientCertificate)

stephentoub commented on src/System.Net.HttpListener/src/System/Net/HttpListenerRequest.Windows.cs:
```diff
@@ -0,0 +1,1246 @@
+                    else
+                    {
+                        header = header.ToLower(CultureInfo.InvariantCulture);
+                        _keepAlive = header.IndexOf("close") < 0 || header.IndexOf("keep-alive") >= 0;
```
Would be better as:
```C#
_keepAlive =
    header.IndexOf("close", StringComparison.InvariantCultureIgnoreCase) < 0 ||
    header.IndexOf("keep-alive", StringComparison.InvariantCultureIgnoreCase) >= 0;
```

Priya91 replied:
done.

stephentoub commented on src/System.Net.HttpListener/src/System/Net/HttpListenerRequest.Windows.cs:
```diff
@@ -0,0 +1,1246 @@
+                }
+
+                //GlobalLog.Print("HttpListenerRequest#" + LoggingHash.HashString(this) + "::KeepAlive_get() returning:" + m_KeepAlive);
+                return _keepAlive == true;
```
>return _keepAlive == true; [](start = 16, length = 26)

Nit: ```return _keepAlive;```

Priya91 replied:
done.

stephentoub commented on src/System.Net.HttpListener/src/System/Net/HttpListenerRequest.Windows.cs:
```diff
@@ -0,0 +1,1246 @@
+                    if (asyncResult != null)
+                    {
+                        asyncResult.InternalCleanup();
+                    }
```
Nit: ```asyncResult?.InternalCleanup();```

Priya91 replied:
done.

stephentoub commented on src/System.Net.HttpListener/src/System/Net/HttpListenerRequestContext.cs:
```diff
@@ -0,0 +1,32 @@
+            return _request.GetChannelBinding();
+        }
+
+        private HttpListenerRequest _request;
```
Nit: should be moved to the top of the class, and it looks like it can be readonly

stephentoub commented on src/System.Net.HttpListener/src/System/Net/HttpListenerRequestUriBuilder.cs:
```diff
@@ -0,0 +1,479 @@
+        {
+            s_utf8Encoding = new UTF8Encoding(false, true);
+            s_ansiEncoding = Encoding.GetEncoding(0, new EncoderExceptionFallback(), new DecoderExceptionFallback());
+        }
```
It'd be better to initialize these where the fields are declared and remove the explicit cctor.

Priya91 replied:
done.

stephentoub commented on src/System.Net.HttpListener/src/System/Net/HttpListenerRequestUriBuilder.cs:
```diff
@@ -0,0 +1,479 @@
+            else
+            {
+                _cookedUriQuery = cookedUriQuery;
+            }
```
Nit: ```_cookedUriQuery = cookedUriQuery ?? string.Empty;```

stephentoub commented on src/System.Net.HttpListener/src/System/Net/HttpListenerRequestUriBuilder.cs:
```diff
@@ -0,0 +1,479 @@
+                }
+                else
+                {
+                    octetString.Append(" ");
```
>" " [](start = 39, length = 3)

Nit: ' '

Priya91 replied:
done

stephentoub commented on src/System.Net.HttpListener/src/System/Net/HttpListenerResponse.Windows.cs:
```diff
@@ -0,0 +1,1018 @@
+        }
+
+        // We MUST NOT send message-body when we send responses with these Status codes
+        private static readonly int[] s_NoResponseBody = { 100, 101, 204, 205, 304 };
```
>s_N [](start = 38, length = 3)

Nit: s_noResponseBody

stephentoub commented on src/System.Net.HttpListener/src/System/Net/HttpListenerResponse.Windows.cs:
```diff
@@ -0,0 +1,1018 @@
+        {
+            if (cookie == null)
+            {
+                throw new ArgumentNullException("cookie");
```
nameof(cookie)

stephentoub commented on src/System.Net.HttpListener/src/System/Net/HttpListenerResponse.Windows.cs:
```diff
@@ -0,0 +1,1018 @@
+        {
+            if (cookie == null)
+            {
+                throw new ArgumentNullException("cookie");
```
nameof(cookie)

stephentoub commented on src/System.Net.HttpListener/src/System/Net/HttpListenerResponse.Windows.cs:
```diff
@@ -0,0 +1,1018 @@
+                }
+                else
+                {
+                    throw new ArgumentOutOfRangeException("value", SR.net_clsmall);
```
nameof(value)

stephentoub commented on src/System.Net.HttpListener/src/System/Net/HttpListenerResponse.Windows.cs:
```diff
@@ -0,0 +1,1018 @@
+                CheckDisposed();
+                if (value == null)
+                {
+                    throw new ArgumentNullException("value");
```
nameof(value), here and below

stephentoub commented on src/System.Net.HttpListener/src/System/Net/HttpListenerResponse.Windows.cs:
```diff
@@ -0,0 +1,1018 @@
+                CheckDisposed();
+                if (responseEntity == null)
+                {
+                    throw new ArgumentNullException("responseEntity");
```
nameof(responseEntity)

stephentoub commented on src/System.Net.HttpListener/src/System/Net/HttpListenerTimeoutManager.cs:
```diff
@@ -0,0 +1,10 @@
+{
+    public partial class HttpListenerTimeoutManager
+    {
+    }
```
Is the idea here that common stuff from the Windows and eventual Unix implementations will be factored out to here?  Otherwise this isn't needed and the type doesn't need to be partial.

stephentoub commented on src/System.Net.HttpListener/src/System/Net/HttpListenerTimeoutManager.Unix.cs:
```diff
@@ -0,0 +1,81 @@
+// See the LICENSE file in the project root for more information.
+
+namespace System.Net
+{
```
Please add a TODO comment

stephentoub commented on src/System.Net.HttpListener/src/System/Net/HttpListenerTimeoutManager.Windows.cs:
```diff
@@ -0,0 +1,194 @@
+
+            if (timeoutValue < 0 || timeoutValue > ushort.MaxValue)
+            {
+                throw new ArgumentOutOfRangeException("value");
```
nameof(value)

stephentoub commented on src/System.Net.HttpListener/src/System/Net/HttpListenerTimeoutManager.Windows.cs:
```diff
@@ -0,0 +1,194 @@
+                //
+                if (value < 0 || value > uint.MaxValue)
+                {
+                    throw new ArgumentOutOfRangeException("value");
```
nameof(value)

stephentoub commented on src/System.Net.HttpListener/src/System/Net/HttpRequestStream.cs:
```diff
@@ -0,0 +1,489 @@
+{
+    internal unsafe class HttpRequestStream : Stream
+    {
+        private HttpListenerContext _httpContext;
```
readonly?

stephentoub commented on src/System.Net.HttpListener/src/System/Net/HttpRequestStream.cs:
```diff
@@ -0,0 +1,489 @@
+
+namespace System.Net
+{
+    internal unsafe class HttpRequestStream : Stream
```
sealed?

stephentoub commented on src/System.Net.HttpListener/src/System/Net/HttpRequestStream.cs:
```diff
@@ -0,0 +1,489 @@
+            }
+            if (size < 0 || size > buffer.Length - offset)
+            {
+                throw new ArgumentOutOfRangeException("size");
```
nameofs

stephentoub commented on src/System.Net.HttpListener/src/System/Net/HttpRequestStream.cs:
```diff
@@ -0,0 +1,489 @@
+            }
+            if (castedAsyncResult.EndCalled)
+            {
+                throw new InvalidOperationException(SR.Format(SR.net_io_invalidendcall, "EndRead"));
```
nameofs

stephentoub commented on src/System.Net.HttpListener/src/System/Net/HttpRequestStream.cs:
```diff
@@ -0,0 +1,489 @@
+            private ThreadPoolBoundHandle _boundHandle;
+            internal NativeOverlapped* m_pOverlapped;
+            internal void* m_pPinnedBuffer;
+            internal uint m_dataAlreadyRead = 0;
```
Nit: _ instead of m_

stephentoub commented on src/System.Net.HttpListener/src/System/Net/HttpResponseStream.cs:
```diff
@@ -0,0 +1,548 @@
+
+namespace System.Net
+{
+    internal unsafe class HttpResponseStream : Stream
```
sealed?

stephentoub commented on src/System.Net.HttpListener/src/System/Net/HttpResponseStream.cs:
```diff
@@ -0,0 +1,548 @@
+namespace System.Net
+{
+    internal unsafe class HttpResponseStream : Stream
+    {
```
readonly?

stephentoub commented on src/System.Net.HttpListener/src/System/Net/HttpResponseStream.cs:
```diff
@@ -0,0 +1,548 @@
+                        {
+                            // TODO:
+                            // here we need some heuristics, some time it is definitely better to split this in 3 write calls
+                            // but for small writes it is probably good enough to just copy the data internally.
```
Is this TODO still valid?

stephentoub commented on src/System.Net.HttpListener/src/System/Net/HttpResponseStream.cs:
```diff
@@ -0,0 +1,548 @@
+                {
+                    // free unmanaged buffer
+                    bufferAsIntPtr.Close();
+                }
```
Nit: ```bufferAsIntPtr?.Close(); // free unmanaged buffer```

stephentoub commented on src/System.Net.HttpListener/src/System/Net/HttpServerSessionHandle.cs:
```diff
@@ -0,0 +1,40 @@
+    //
+    internal sealed class HttpServerSessionHandle : SafeHandleZeroOrMinusOneIsInvalid
+    {
+        private ulong _serverSessionId;
```
readonly?

stephentoub commented on src/System.Net.HttpListener/src/System/Net/ListenerAsyncResult.cs:
```diff
@@ -0,0 +1,167 @@
+
+namespace System.Net
+{
+    internal unsafe class ListenerAsyncResult : LazyAsyncResult
```
sealed?

stephentoub commented on src/System.Net.HttpListener/src/System/Net/ListenerAsyncResult.cs:
```diff
@@ -0,0 +1,167 @@
+            }
+        }
+
+        private AsyncRequestContext _requestContext;
```
readonly?

stephentoub commented on src/System.Net.HttpListener/src/System/Net/WebSockets/HttpListenerWebSocketContext.cs:
```diff
@@ -0,0 +1,136 @@
+        private string _origin;
+        private IEnumerable<string> _secWebSocketProtocols;
+        private string _secWebSocketVersion;
+        private string _secWebSocketKey;
```
Looks like a bunch of these could be readonly

stephentoub commented on src/System.Net.HttpListener/src/System/Net/WebSockets/SafeWebSocketHandle.cs:
```diff
@@ -0,0 +1,29 @@
+            if (this.IsInvalid)
+            {
+                return true;
+            }
```
This should not be necessary

stephentoub commented on src/System.Net.HttpListener/src/System/Net/WebSockets/ServerWebSocket.cs:
```diff
@@ -0,0 +1,86 @@
+
+            WebSocketHelpers.ValidateInnerStream(innerStream);
+            WebSocketHelpers.ValidateOptions(subProtocol, receiveBufferSize, WebSocketBuffer.MinSendBufferSize, keepAliveInterval);
+            WebSocketHelpers.ValidateArraySegment<byte>(internalBuffer, "internalBuffer");
```
nameof(internalBuffer)

stephentoub commented:
>     internal abstract class WebSocketBase : WebSocket, IDisposable

Can we just reuse ManagedWebSocket here?
[https://github.com/dotnet/corefx/blob/master/src/System.Net.WebSockets.Client/src/System/Net/WebSockets/ManagedWebSocket.cs](https://github.com/dotnet/corefx/blob/master/src/System.Net.WebSockets.Client/src/System/Net/WebSockets/ManagedWebSocket.cs)

---
Refers to: src/System.Net.HttpListener/src/System/Net/WebSockets/WebSocketBase.cs:26 in 450dc0f. [](commit_id = 450dc0fb2cb255577101545399f698ed9e6b44c4, deletion_comment = False)

stephentoub commented on src/System.Net.HttpListener/src/System/Net/WebSockets/WebSocketHelpers.cs:
```diff
@@ -0,0 +1,569 @@
+
+            if (count < 0 || count > (buffer.Length - offset))
+            {
+                throw new ArgumentOutOfRangeException("count");
```
nameofs

stephentoub commented on src/System.Net.HttpListener/src/System/Net/WebSockets/WebSocketHelpers.cs:
```diff
@@ -0,0 +1,569 @@
+        {
+            if (string.IsNullOrWhiteSpace(subProtocol))
+            {
+                throw new ArgumentException(SR.net_WebSockets_InvalidEmptySubProtocol, "subProtocol");
```
nameof

stephentoub commented on src/System.Net.HttpListener/src/System/Net/WebSockets/WebSocketHttpListenerDuplexStream.cs:
```diff
@@ -0,0 +1,1238 @@
+
+namespace System.Net.WebSockets
+{
+    internal class WebSocketHttpListenerDuplexStream : Stream, WebSocketBase.IWebSocketStream
```
sealed

stephentoub commented on src/System.Net.HttpListener/tests/SimpleHttpTest.cs:
```diff
@@ -0,0 +1,59 @@
+namespace System.Net.Tests
+{
+    [PlatformSpecific(TestPlatforms.Windows)]
+    public class SimpleHttpTest
```
Obviously we need many more tests.  What's the plan for that?  As part of this PR?  Immediately after?

CIPop replied:
+1 Would be great to have a tracking issue with remaining work and maybe a TODO here to link back to the issue.

Priya91 replied:
Yeah will add tests later, trying to get Eric's changes in first.

stephentoub commented on src/System.Net.Requests/ref/System.Net.Requests.cs:
```diff
@@ -128,6 +128,7 @@ public partial class HttpWebRequest : System.Net.WebRequest
         internal HttpWebRequest() { }
         public string Accept { get { throw null; } set { } }
         public virtual bool AllowReadStreamBuffering { get { throw null; } set { } }
+        public string Connection { get { throw null; } set { } }
```
>public string Connection { get { throw null; } set { } } [](start = 8, length = 56)

Why as part of this PR?  It's also being added in this PR:
[https://github.com/dotnet/corefx/pull/13441](https://github.com/dotnet/corefx/pull/13441)

Priya91 replied:
Hmm, not sure, this is from eric's change. Maybe both did the changes.

stephentoub commented:
:clock1:

CIPop commented:
@Priya91 @karelz @davidsh Have you considered splitting the WebSocket part into S.N.WebSockets.Server instead of bundling it with HttpListener?
Also, may not be obvious but none of that code will work on Win7 as it relies on http.sys opaque upgrades and the WebSocket.dll Win8+ native dll.

CIPop commented on src/System.Net.HttpListener/src/System/Net/WebSockets/WebSocketProtocolComponent.cs:
```diff
@@ -0,0 +1,709 @@
+
+namespace System.Net.WebSockets
+{
+    internal static class WebSocketProtocolComponent
```
This should be in Interop, not src.

CIPop commented on src/System.Net.HttpListener/src/System/Net/WebSockets/WebSocketProtocolComponent.cs:
```diff
@@ -0,0 +1,709 @@
+            }
+        }
+
+        [DllImport(Interop.Libraries.WebSocket, EntryPoint = "WebSocketCreateClientHandle", ExactSpelling = true)]
```
This needs some extra work: the Interop layer needs to be properly split into separate files, etc.

CIPop commented on src/System.Net.HttpListener/src/System/Net/WebSockets/WebSocketProtocolComponent.cs:
```diff
@@ -0,0 +1,709 @@
+                throw new AccessViolationException();
+            }
+
+            // structure of HttpHeader:
```
This is a bad way to emulate a struct: we risk missing padding and corrupting memory on other architectures.

CIPop commented on src/System.Net.HttpListener/src/System/Net/WebSockets/WebSocketProtocolComponent.cs:
```diff
@@ -0,0 +1,709 @@
+
+            HttpHeader[] httpHeaders = new HttpHeader[nativeHeaderCount];
+
+            // structure of HttpHeader:
```
Same as above. We should at least open bugs / TODOs.

stephentoub commented:
> Also, may not be obvious but none of that code will work on Win7 as it relies on http.sys opaque upgrades and the WebSocket.dll Win8+ native dll.

My assumption has been that we could use whatever managed implementation we port from Mono to enable at least some basic functionality on Win7.  For any real web server use, we really want folks to use ASP.NET, kestrel, etc., so at least personally I view this HttpListener work as supporting mostly non-production web server scenarios.

CIPop commented:
> My assumption has been that we could use whatever managed implementation we port from Mono

There are two options for this: 

1. use the managed WS server implementation that's used for WebListener (the HttpListener variant in ASP.Net)
2. use the WS server implementation + HttpListener implementation from Mono

If we want 1, and for better componentization, we could work with the ASP.Net team and extract the WS server code in a different contract that can be reused across both implementations.

stephentoub commented:
> use the managed WS server implementation that's used for WebListener

Isn't WebListener based on http.sys?

CIPop commented:
@karelz @stephentoub @davidsh In an offline discussion we've decided that there are some advantages if we split out the WebSockets implementation and add it to System.Net. I've opened #13570 to track this.

Priya91 commented on src/Common/src/Interop/Windows/HttpApi/Interop.HttpApi.cs:
```diff
@@ -0,0 +1,1017 @@
+{
+    internal unsafe static partial class HttpApi
+    {
+        internal static readonly HTTPAPI_VERSION Version = new HTTPAPI_VERSION() { HttpApiMajorVersion = 2, HttpApiMinorVersion = 0 };
```
fixed.

Priya91 commented on src/Common/src/Interop/Windows/HttpApi/Interop.HttpApi.cs:
```diff
@@ -0,0 +1,1017 @@
+    {
+        internal static readonly HTTPAPI_VERSION Version = new HTTPAPI_VERSION() { HttpApiMajorVersion = 2, HttpApiMinorVersion = 0 };
+
+        internal static readonly bool Supported = InitHttpApi(Version);
```
fixed.

Priya91 commented on src/Common/src/Interop/Windows/HttpApi/Interop.HttpApi.cs:
```diff
@@ -0,0 +1,1017 @@
+        internal class SafeLocalFreeChannelBinding : ChannelBinding
+        {
+            private const int LMEM_FIXED = 0;
+            private int _size;
```
fixed.

Priya91 commented on src/Common/src/Interop/Windows/HttpApi/Interop.HttpApi.cs:
```diff
@@ -0,0 +1,1017 @@
+                "WWW-Authenticate",
+            };
+
+            private static Dictionary<string, int> s_hashtable = CreateTable();
```
fixed.

clrjunkie commented:
Big Plus for working on this!
Big Minus for "viewing this work as supporting mostly non-production web server scenarios." 

HttpListener must be treated  as a key infrastructure component for productions systems from day one.

It only stands to reason that this would ultimately be the underlying http server for ASP.Net Core (as in the case of Full FX ASP.NET self hosting ) and "Kestrel" would be a thing of the past...

stephentoub commented:
> HttpListener must be treated as a key infrastructure component for productions systems from day one. It only stands to reason that this would ultimately be the underlying http server for ASP.Net Core (as in the case of Full FX ASP.NET self hosting ) and "Kestrel" would be a thing of the past

Thank you for the feedback, but no.  The ASP.NET team (both at Microsoft and the community) has and continues to invest years and years of dev time into making that stack as efficient and programmable as possible, on both Windows and Unix, evolving it both in terms of implementation and API to do so.  We do not plan to duplicate such efforts with HttpListener, nor are we able to take the API-level breaks necessary to bring it up to performance par. It is being exposed as part of corefx for compatibility, to enable code currently using it, but for production scenarios, I recommend using kestrel, ASP.NET, etc.

clrjunkie commented:
@stephentoub The "ASP.NET" Stack has nothing to do with either HttpListener or "Kestrel" and if backward compatibility drives this effort then your duplicating efforts (and wasting time) as we speak... since nobody should care about the underlying HttpListener implementation (e.g Http.sys) but rather about the *interface* so for all intent and purposes "Kestrel" should expose  HttpListener interface and get it over with. It's a simple interface and there is no reason it should not suit production scenarios regardless of the stack built on top of it.

stephentoub commented:
That is not the direction we are going. Thanks.

Priya91 commented on src/System.Net.HttpListener/src/System/Net/HttpListener.Unix.cs:
```diff
@@ -0,0 +1,216 @@
+namespace System.Net
+{
+    public sealed unsafe partial class HttpListener
+    {
```
fixed.

CIPop commented on src/Common/src/Interop/Windows/HttpApi/Interop.HttpApi.cs:
```diff
@@ -0,0 +1,1017 @@
+
+        private static bool InitHttpApi(HTTPAPI_VERSION version)
+        {
+            uint statusCode = HttpInitialize(version, (uint)HTTP_FLAGS.HTTP_INITIALIZE_SERVER, null);
```
@Priya91 To clarify: this is good as written and no more changes are required. CoreFX will only run on Win7 and above which support the above initialization code.

Priya91 commented on src/System.Net.HttpListener/src/System/Net/HttpListener.Windows.cs:
```diff
@@ -0,0 +1,2135 @@
+            if (!disposing)
+            {
+                return;
+            }
```
Don't see a destructor. Will remove

Priya91 commented on src/System.Net.HttpListener/src/System/Net/HttpListener.Windows.cs:
```diff
@@ -0,0 +1,2135 @@
+                }
+                if (castedAsyncResult.EndCalled)
+                {
+                    throw new InvalidOperationException(SR.Format(SR.net_io_invalidendcall, "EndGetContext"));
```
fixed.

Priya91 commented on src/System.Net.HttpListener/src/System/Net/HttpListenerContext.Windows.cs:
```diff
@@ -0,0 +1,264 @@
+        private AuthenticationSchemes _authenticationSchemes;
+        private ExtendedProtectionPolicy _extendedProtectionPolicy;
+
+        internal const string NTLM = "NTLM";
```
Done.

Priya91 commented on src/System.Net.HttpListener/src/System/Net/HttpListenerException.cs:
```diff
@@ -0,0 +1,28 @@
+        {
+            //GlobalLog.Print("HttpListenerException::.ctor(int) " + NativeErrorCode.ToString() + ":" + Message);
+        }
+    }
```
done.

Priya91 commented on src/System.Net.HttpListener/src/System/Net/HttpListenerRequest.Windows.cs:
```diff
@@ -0,0 +1,1246 @@
+
+        public Task<X509Certificate2> GetClientCertificateAsync()
+        {
+            return Task<X509Certificate2>.Factory.FromAsync(BeginGetClientCertificate, EndGetClientCertificate, null);
```
done.

Priya91 commented on src/System.Net.HttpListener/src/System/Net/HttpListenerRequestUriBuilder.cs:
```diff
@@ -0,0 +1,479 @@
+            int codePointValue;
+            if (!int.TryParse(codePoint, NumberStyles.HexNumber, null, out codePointValue))
+            {
+                LogWarning("AppendUnicodeCodePointValuePercentEncoded",
```
done.

Priya91 commented on src/System.Net.HttpListener/src/System/Net/HttpListenerResponse.Unix.cs:
```diff
@@ -0,0 +1,197 @@
+using System.Text;
+
+namespace System.Net
+{
```
done

Priya91 commented on src/System.Net.HttpListener/src/System.Net.HttpListener.csproj:
```diff
@@ -0,0 +1,152 @@
+  </ItemGroup>
+  <ItemGroup>
+    <!-- Temporary package reference until NuGet packages are updated -->
+    <ProjectReference Include="..\..\System.Net.Requests\pkg\System.Net.Requests.pkgproj" />
```
It's used in WebSocketHelpers.

Priya91 replied:
The calling api, was not referenced anywhere. Removed this.

Priya91 commented on src/System.Net.HttpListener/src/System/Net/WebSockets/WebSocketBuffer.cs:
```diff
@@ -0,0 +1,694 @@
+            Debug.Assert(false, "'buffer' MUST reference a memory segment within the pinned InternalBuffer.");
+            // Indicates a violation in the contract with native Websocket.dll and could indicate 
+            // memory corruption because the internal buffer is shared between managed and native code
+            throw new AccessViolationException();
```
This is a behavior change from desktop, and not a nit. Can you elaborate more on what exception should be thrown here? Please file issue

Priya91 commented:
@stephentoub @CIPop Addressed the PR feedback and filed issues on pending workitems. Plan to merge PR after CI passes, will continue watching the PR, and address additional comments in separate PR. This PR is getting too big to keep track of the work.

stephentoub commented:
Thanks, @Priya91.
